In [261]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

Importing the Dataset

In [262]:
data = pd.read_csv('bank_Churn_dataset.csv')
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

Dropping the Unnecessary Columns

In [263]:
data.drop(['RowNumber', 'CustomerId', 'Surname'], axis = 1, inplace = True)
data.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42       2       0.00              1   
1          608     Spain  Female   41       1   83807.86              1   
2          502    France  Female   42       8  159660.80              3   
3          699    France  Female   39       1       0.00              2   
4          850     Spain  Female   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

In [264]:
data.dtypes

CreditScore          int64
Geography           object
Gender              object
Age                  int64
Tenure               int64
Balance            float64
NumOfProducts        int64
HasCrCard            int64
IsActiveMember       int64
EstimatedSalary    float64
Exited               int64
dtype: object

In [265]:
data.isnull().sum()

CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64

Seeing for if  there is any unique dataset in Object dtatype to convert in to Numeric Form

In [266]:
for column in data:
    if data[column].dtypes == 'object':
        print(data[column].unique())

['France' 'Spain' 'Germany']
['Female' 'Male']


Replacing the categorical data to its numeric form

In [267]:
data['Gender'].replace({'Female':1, 'Male':0}, inplace = True)
data['Geography'].replace({'France':0, 'Spain':1, 'Germany':2}, inplace = True)

In [268]:
data.head()

CreditScore  Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619          0       1   42       2       0.00              1   
1          608          1       1   41       1   83807.86              1   
2          502          0       1   42       8  159660.80              3   
3          699          0       1   39       1       0.00              2   
4          850          1       1   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

In [269]:
for column in data:
    if data[column].dtypes == 'object':
        print(data[column].unique())
#there are no unique elements

Scaling the data for the columns ['CreditScore', 'Age', 'Tenure', 'Balance', 'EstimatedSalary']

In [270]:
data.head()

CreditScore  Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619          0       1   42       2       0.00              1   
1          608          1       1   41       1   83807.86              1   
2          502          0       1   42       8  159660.80              3   
3          699          0       1   39       1       0.00              2   
4          850          1       1   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

In [271]:
from sklearn import preprocessing
scaler = preprocessing.MinMaxScaler()
scaling_column = ['CreditScore', 'Age', 'Tenure', 'Balance', 'EstimatedSalary']
data[scaling_column] = scaler.fit_transform(data[scaling_column])

In [272]:
data.head()

CreditScore  Geography  Gender       Age  Tenure   Balance  NumOfProducts  \
0        0.538          0       1  0.324324     0.2  0.000000              1   
1        0.516          1       1  0.310811     0.1  0.334031              1   
2        0.304          0       1  0.324324     0.8  0.636357              3   
3        0.698          0       1  0.283784     0.1  0.000000              2   
4        1.000          1       1  0.337838     0.2  0.500246              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1         0.506735       1  
1          0               1         0.562709       0  
2          1               0         0.569654       1  
3          0               0         0.469120       0  
4          1               1         0.395400       0

Allocating the X dataset and y dataset

In [273]:
X = data.drop('Exited', axis = 1)
y = data['Exited']

Splitting the dataset with 80% training model and 20 % testing Model

In [274]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 43)

In [275]:
X_train.shape, X_test.shape

((8000, 10), (2000, 10))

In [276]:
import tensorflow
from tensorflow import keras

Creating the Artificial Neural Network Model using mini btch gradient Descent

 I have taken mini batch size as 10 because to get high accuracy evenhough it takes time

In [277]:
model = keras.Sequential([
    keras.layers.Dense(15, input_shape = (10, ), activation = 'relu'),
    keras.layers.Dense(10, activation = 'relu'),
    keras.layers.Dense(1,activation='sigmoid')
])

model.compile(
    optimizer = 'adam',
    loss = 'binary_crossentropy', 
    metrics = ['accuracy']
)

model.fit(X_train, y_train, epochs=100, batch_size = 10)

Epoch 1/100
8000/8000 [==============================] - 2s 232us/sample - loss: 0.4896 - acc: 0.7950
Epoch 2/100
8000/8000 [==============================] - 2s 211us/sample - loss: 0.4509 - acc: 0.8067
Epoch 3/100
8000/8000 [==============================] - 2s 208us/sample - loss: 0.4288 - acc: 0.8146
Epoch 4/100
8000/8000 [==============================] - 2s 213us/sample - loss: 0.4101 - acc: 0.8219
Epoch 5/100
8000/8000 [==============================] - 2s 210us/sample - loss: 0.3958 - acc: 0.8325
Epoch 6/100
8000/8000 [==============================] - 2s 232us/sample - loss: 0.3850 - acc: 0.8354
Epoch 7/100
8000/8000 [==============================] - 2s 243us/sample - loss: 0.3768 - acc: 0.8421
Epoch 8/100
8000/8000 [==============================] - 2s 214us/sample - loss: 0.3690 - acc: 0.8446
Epoch 9/100
8000/8000 [==============================] - 2s 220us/sample - loss: 0.3621 - acc: 0.8485
Epoch 10/100
8000/8000 [==============================] - 2s 206us/sample - loss: 

In [278]:
model.evaluate(X_test, y_test)

2000/2000 [==============================] - 0s 104us/sample - loss: 0.3496 - acc: 0.8590


[0.34961445689201354, 0.859]

In [279]:
y_predicted = model.predict(X_test)
y_predicted[:5]

array([[0.24859753],
       [0.1005443 ],
       [0.17100799],
       [0.18345013],
       [0.24051797]], dtype=float32)

In [280]:
y_test[:5]

9415    0
6377    1
8019    0
7754    1
4961    0
Name: Exited, dtype: int64

In [281]:
y_predicted = np.asanyarray(y_predicted)
y_predicted

array([[0.24859753],
       [0.1005443 ],
       [0.17100799],
       ...,
       [0.00647864],
       [0.42368752],
       [0.07129672]], dtype=float32)

In [282]:
len(y_predicted)

2000

In [283]:
y_predicted[1]

array([0.1005443], dtype=float32)

In [284]:
y_predict = []
for i in range(len(y_predicted)):
    if y_predicted[i] > 0.5:
        y_predict.append(1)
    else:
        y_predict.append(0)


In [285]:
y_predict[:5]

[0, 0, 0, 0, 0]

Confusion matrix

In [286]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_predict)

array([[1518,   70],
       [ 212,  200]], dtype=int64)

Classification Report

In [287]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_predict))

              precision    recall  f1-score   support

           0       0.88      0.96      0.92      1588
           1       0.74      0.49      0.59       412

    accuracy                           0.86      2000
   macro avg       0.81      0.72      0.75      2000
weighted avg       0.85      0.86      0.85      2000



# Handling imbalanced data using Undersampling

In [288]:
data['Exited'].value_counts()

0    7963
1    2037
Name: Exited, dtype: int64

In [289]:
class_0_counts, class_1_counts = data['Exited'].value_counts()
print(class_0_counts, class_1_counts)

7963 2037


In [290]:
class_0_data = data[data['Exited'] == 0] #class 0 samples
class_1_data = data[data['Exited'] == 1] #class 1 samples

In [291]:
class_0_data.shape, class_1_data.shape

((7963, 11), (2037, 11))

Undersampling the 0 class dataset

In [292]:
under_sampled_class_0 = class_0_data.sample(class_1_counts)
under_sampled_class_0.shape

(2037, 11)

In [293]:
print(under_sampled_class_0.shape, class_1_data.shape)
# both the dataset has same shape

(2037, 11) (2037, 11)


Concating both the datasets

In [294]:
under_sample_dataset = pd.concat([under_sampled_class_0, class_1_data], axis = 0)
under_sample_dataset.shape

(4074, 11)

In [295]:
X = under_sample_dataset.drop('Exited', axis = 1)
y = under_sample_dataset['Exited']

Splitting and training the model using ANN

In [296]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 43)

model = keras.Sequential([
    keras.layers.Dense(15, input_shape = (10, ), activation = 'relu'),
    keras.layers.Dense(10, activation = 'relu'),
    keras.layers.Dense(1,activation='sigmoid')
])

model.compile(
    optimizer = 'adam',
    loss = 'binary_crossentropy', 
    metrics = ['accuracy']
)

model.fit(X_train, y_train, epochs=100, batch_size = 10)

Epoch 1/100
3259/3259 [==============================] - 1s 280us/sample - loss: 0.6625 - acc: 0.5956
Epoch 2/100
3259/3259 [==============================] - 1s 215us/sample - loss: 0.6259 - acc: 0.6655
Epoch 3/100
3259/3259 [==============================] - 1s 268us/sample - loss: 0.6024 - acc: 0.6809
Epoch 4/100
3259/3259 [==============================] - 1s 270us/sample - loss: 0.5828 - acc: 0.6990
Epoch 5/100
3259/3259 [==============================] - 1s 221us/sample - loss: 0.5705 - acc: 0.7060
Epoch 6/100
3259/3259 [==============================] - 1s 219us/sample - loss: 0.5567 - acc: 0.7162
Epoch 7/100
3259/3259 [==============================] - 1s 231us/sample - loss: 0.5476 - acc: 0.7232
Epoch 8/100
3259/3259 [==============================] - 1s 262us/sample - loss: 0.5379 - acc: 0.7352
Epoch 9/100
3259/3259 [==============================] - 1s 222us/sample - loss: 0.5315 - acc: 0.7334
Epoch 10/100
3259/3259 [==============================] - 1s 212us/sample - loss: 

Evaluating the Model

In [297]:
model.evaluate(X_test, y_test)

815/815 [==============================] - 0s 192us/sample - loss: 0.4732 - acc: 0.7632


[0.47321462346000903, 0.7631902]

Predicting the model and converting the output as 1 and 0 using 0.5 as threshold

In [298]:
y_predicted = model.predict(X_test)
y_predicted = np.asanyarray(y_predicted)
y_predict = []
for i in range(len(y_predicted)):
    if y_predicted[i] > 0.5:
        y_predict.append(1)
    else:
        y_predict.append(0)

Confusion Matrix

In [299]:
confusion_matrix(y_test, y_predict)

array([[328,  75],
       [118, 294]], dtype=int64)

Classification report

In [300]:
print(classification_report(y_test, y_predict))

              precision    recall  f1-score   support

           0       0.74      0.81      0.77       403
           1       0.80      0.71      0.75       412

    accuracy                           0.76       815
   macro avg       0.77      0.76      0.76       815
weighted avg       0.77      0.76      0.76       815



Hence the f1-score of  the classes 0 - (0.92) and 1 - (0.59) has changed to classes 0 - (0.77) and 1 - (0.75)..  
So, overall f1 Score is better than the previous ANN model wihout handling imbalanced dataset

#  Handling imbalanced data using Oversampling : Duplicate Copy

In [301]:
class_0_counts, class_1_counts

(7963, 2037)

In [302]:
class_0_data.shape, class_1_data.shape

((7963, 11), (2037, 11))

Making the dataset equal in number randomly copying any dataset in class1

In [303]:
under_sampled_class_1 = class_1_data.sample(class_0_counts, replace=True)
class_0_data.shape, under_sampled_class_1.shape

((7963, 11), (7963, 11))

Concatinating both the dataset

In [304]:
over_sample_dataset = pd.concat([class_0_data, under_sampled_class_1], axis = 0)
over_sample_dataset.shape

(15926, 11)

In [305]:
X = over_sample_dataset.drop('Exited', axis = 1)
y = over_sample_dataset['Exited']

Splitting the in train and test and executing the ANN Model

In [306]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 43)

model = keras.Sequential([
    keras.layers.Dense(15, input_shape = (10, ), activation = 'relu'),
    keras.layers.Dense(10, activation = 'relu'),
    keras.layers.Dense(1,activation='sigmoid')
])

model.compile(
    optimizer = 'adam',
    loss = 'binary_crossentropy', 
    metrics = ['accuracy']
)

model.fit(X_train, y_train, epochs=100, batch_size = 10)

Epoch 1/100
12740/12740 [==============================] - 3s 231us/sample - loss: 0.6199 - acc: 0.6571
Epoch 2/100
12740/12740 [==============================] - 3s 244us/sample - loss: 0.5698 - acc: 0.7049
Epoch 3/100
12740/12740 [==============================] - 3s 229us/sample - loss: 0.5409 - acc: 0.7254
Epoch 4/100
12740/12740 [==============================] - 3s 210us/sample - loss: 0.5216 - acc: 0.7418
Epoch 5/100
12740/12740 [==============================] - 3s 240us/sample - loss: 0.5103 - acc: 0.7451
Epoch 6/100
12740/12740 [==============================] - 3s 209us/sample - loss: 0.5008 - acc: 0.7513
Epoch 7/100
12740/12740 [==============================] - 3s 211us/sample - loss: 0.4940 - acc: 0.7546
Epoch 8/100
12740/12740 [==============================] - 3s 230us/sample - loss: 0.4894 - acc: 0.7554
Epoch 9/100
12740/12740 [==============================] - 3s 218us/sample - loss: 0.4862 - acc: 0.7596
Epoch 10/100
12740/12740 [==============================] - 3s 2

Evaluating the Model

In [307]:
model.evaluate(X_test, y_test)

3186/3186 [==============================] - 0s 80us/sample - loss: 0.4359 - acc: 0.7916


[0.43592421715168256, 0.7915882]

Predicting the model and converting the output as 1 and 0 using 0.5 as threshold

In [308]:
y_predicted = model.predict(X_test)
y_predicted = np.asanyarray(y_predicted)
y_predict = []
for i in range(len(y_predicted)):
    if y_predicted[i] > 0.5:
        y_predict.append(1)
    else:
        y_predict.append(0)

Confusion Matrix

In [309]:
confusion_matrix(y_test, y_predict)

array([[1294,  290],
       [ 374, 1228]], dtype=int64)

Classification Report

In [310]:
print(classification_report(y_test, y_predict))

              precision    recall  f1-score   support

           0       0.78      0.82      0.80      1584
           1       0.81      0.77      0.79      1602

    accuracy                           0.79      3186
   macro avg       0.79      0.79      0.79      3186
weighted avg       0.79      0.79      0.79      3186



Hence the f1-score of  the classes 0 - (0.92) and 1 - (0.59) has changed to classes 0 - (0.80) and 1 - (0.79)..  
So, overall f1 Score is better than the previous ANN model wihout handling imbalanced dataset

#  Handling imbalanced data using Oversampling : SMOTE

In [311]:
class_0_counts, class_1_counts

(7963, 2037)

Splitting X and y data

In [312]:
X = data.drop('Exited', axis = 1)
y = data['Exited']

SMOTE uses KNN to make duplicate dataset which is better than randomly copying the data 

In [313]:
from imblearn.over_sampling import SMOTE

In [314]:
smote = SMOTE(sampling_strategy='minority')
X_data, y_data = smote.fit_sample(X, y)

In [315]:
y_data.value_counts()
# both the classes data is same

1    7963
0    7963
Name: Exited, dtype: int64

Splitting the train and test sample and executing the ANN model

In [316]:
X_train, X_test, y_train, y_test = train_test_split(X_data, y_data, test_size = 0.2, random_state = 43)

model = keras.Sequential([
    keras.layers.Dense(15, input_shape = (10, ), activation = 'relu'),
    keras.layers.Dense(10, activation = 'relu'),
    keras.layers.Dense(1,activation='sigmoid')
])

model.compile(
    optimizer = 'adam',
    loss = 'binary_crossentropy', 
    metrics = ['accuracy']
)

model.fit(X_train, y_train, epochs=100, batch_size = 10)

Epoch 1/100
12740/12740 [==============================] - 3s 240us/sample - loss: 0.6176 - acc: 0.6643
Epoch 2/100
12740/12740 [==============================] - 3s 248us/sample - loss: 0.5516 - acc: 0.7189
Epoch 3/100
12740/12740 [==============================] - 3s 261us/sample - loss: 0.5202 - acc: 0.7453
Epoch 4/100
12740/12740 [==============================] - 3s 235us/sample - loss: 0.5037 - acc: 0.7564
Epoch 5/100
12740/12740 [==============================] - 3s 221us/sample - loss: 0.4942 - acc: 0.7629
Epoch 6/100
12740/12740 [==============================] - 3s 219us/sample - loss: 0.4887 - acc: 0.7666
Epoch 7/100
12740/12740 [==============================] - 3s 223us/sample - loss: 0.4852 - acc: 0.7640
Epoch 8/100
12740/12740 [==============================] - 3s 248us/sample - loss: 0.4801 - acc: 0.7695
Epoch 9/100
12740/12740 [==============================] - 3s 237us/sample - loss: 0.4769 - acc: 0.7690
Epoch 10/100
12740/12740 [==============================] - 3s 2

Evaluating the Model

In [317]:
model.evaluate(X_test, y_test)

3186/3186 [==============================] - 0s 88us/sample - loss: 0.4299 - acc: 0.7982


[0.42990752972302315, 0.7981795]

Predicting the model and converting the output as 1 and 0 using 0.5 as threshold

In [318]:
y_predicted = model.predict(X_test)
y_predicted = np.asanyarray(y_predicted)
y_predict = []
for i in range(len(y_predicted)):
    if y_predicted[i] > 0.5:
        y_predict.append(1)
    else:
        y_predict.append(0)

Confusion Matrix

In [319]:
confusion_matrix(y_test, y_predict)

array([[1302,  295],
       [ 348, 1241]], dtype=int64)

Classification Report

In [320]:
print(classification_report(y_test, y_predict))

              precision    recall  f1-score   support

           0       0.79      0.82      0.80      1597
           1       0.81      0.78      0.79      1589

    accuracy                           0.80      3186
   macro avg       0.80      0.80      0.80      3186
weighted avg       0.80      0.80      0.80      3186



Hence the f1-score of  the classes 0 - (0.92) and 1 - (0.59) has changed to classes 0 - (0.80) and 1 - (0.79)..  
So, overall f1 Score is better than the previous ANN model wihout handling imbalanced dataset

#  Handling imbalanced data using Ensemble

In [321]:
X = data.drop('Exited', axis = 1)
y = data['Exited']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 43)

In [322]:
data['Exited'].value_counts()

0    7963
1    2037
Name: Exited, dtype: int64

In [323]:
class_0_data.shape, class_1_data.shape

((7963, 11), (2037, 11))

In [324]:
print("Actual count", (7963/2037))
print("Round of count", round(7963/2037))

Actual count 3.9091801669121256
Round of count 4


The data is approximately in ratio of 4

Now taking the dataset of class 0 and class 1 and making them equal in examples by dividing the training set in same  
length as test data and creaing different models

Creating the Funtion to make different classes of dataset 

In [325]:
def creating_class_of_same_training_and_testing_dataset(start_index, end_index):
    dataset_ = pd.concat([class_0_data[start_index:end_index], class_1_data], axis = 0)
    X_train = dataset_.drop('Exited', axis = 1)
    y_train = dataset_['Exited']
    return X_train, y_train
    

Creating 4 different datset classes of same number of both same training and testing dataset 

In [326]:
X_train1, y_train1 = creating_class_of_same_training_and_testing_dataset(0,2037)
X_train2, y_train2 = creating_class_of_same_training_and_testing_dataset(2037,4074)
X_train3, y_train3 = creating_class_of_same_training_and_testing_dataset(4074,6111)
X_train4, y_train4 = creating_class_of_same_training_and_testing_dataset(6111,7964)

In [327]:
X_train1.shape, X_test.shape

((4074, 10), (2000, 10))

In [328]:
model = keras.Sequential([
    keras.layers.Dense(15, input_shape = (10, ), activation = 'relu'),
    keras.layers.Dense(10, activation = 'relu'),
    keras.layers.Dense(1,activation='sigmoid')
])

model.compile(
    optimizer = 'adam',
    loss = 'binary_crossentropy', 
    metrics = ['accuracy']
)

model.fit(X_train1, y_train1, epochs=100, batch_size = 10)

Epoch 1/100
4074/4074 [==============================] - 1s 299us/sample - loss: 0.6613 - acc: 0.6073
Epoch 2/100
4074/4074 [==============================] - 1s 258us/sample - loss: 0.6173 - acc: 0.6674
Epoch 3/100
4074/4074 [==============================] - 1s 245us/sample - loss: 0.5909 - acc: 0.6969
Epoch 4/100
4074/4074 [==============================] - 1s 256us/sample - loss: 0.5671 - acc: 0.7106
Epoch 5/100
4074/4074 [==============================] - 1s 295us/sample - loss: 0.5484 - acc: 0.7219
Epoch 6/100
4074/4074 [==============================] - 1s 232us/sample - loss: 0.5338 - acc: 0.7315
Epoch 7/100
4074/4074 [==============================] - 1s 241us/sample - loss: 0.5232 - acc: 0.7435
Epoch 8/100
4074/4074 [==============================] - 1s 270us/sample - loss: 0.5175 - acc: 0.7410
Epoch 9/100
4074/4074 [==============================] - 1s 230us/sample - loss: 0.5129 - acc: 0.7467
Epoch 10/100
4074/4074 [==============================] - 1s 229us/sample - loss: 

In [329]:
model.evaluate(X_test, y_test)

2000/2000 [==============================] - 0s 177us/sample - loss: 0.4275 - acc: 0.8125


[0.4274610075950623, 0.8125]

In [330]:
y_predicted_1 = model.predict(X_test)
y_predicted_1 = np.asanyarray(y_predicted_1)
y_predict_1 = []
for i in range(len(y_predicted_1)):
    if y_predicted_1[i] > 0.5:
        y_predict_1.append(1)
    else:
        y_predict_1.append(0)

In [331]:
print(classification_report(y_test, y_predict_1))

              precision    recall  f1-score   support

           0       0.92      0.84      0.88      1588
           1       0.53      0.70      0.61       412

    accuracy                           0.81      2000
   macro avg       0.73      0.77      0.74      2000
weighted avg       0.84      0.81      0.82      2000



for 2nd class dataset

In [332]:
X_train2.shape, X_test.shape

((4074, 10), (2000, 10))

In [333]:
model = keras.Sequential([
    keras.layers.Dense(15, input_shape = (10, ), activation = 'relu'),
    keras.layers.Dense(10, activation = 'relu'),
    keras.layers.Dense(1,activation='sigmoid')
])

model.compile(
    optimizer = 'adam',
    loss = 'binary_crossentropy', 
    metrics = ['accuracy']
)

model.fit(X_train2, y_train2, epochs=100, batch_size = 10)

Epoch 1/100
4074/4074 [==============================] - 2s 395us/sample - loss: 0.7219 - acc: 0.5712
Epoch 2/100
4074/4074 [==============================] - 1s 313us/sample - loss: 0.6426 - acc: 0.6284
Epoch 3/100
4074/4074 [==============================] - 1s 348us/sample - loss: 0.6217 - acc: 0.6524
Epoch 4/100
4074/4074 [==============================] - 1s 351us/sample - loss: 0.6054 - acc: 0.6647
Epoch 5/100
4074/4074 [==============================] - 1s 331us/sample - loss: 0.5891 - acc: 0.6848
Epoch 6/100
4074/4074 [==============================] - 1s 309us/sample - loss: 0.5756 - acc: 0.6949
Epoch 7/100
4074/4074 [==============================] - 1s 345us/sample - loss: 0.5623 - acc: 0.7089
Epoch 8/100
4074/4074 [==============================] - 1s 332us/sample - loss: 0.5472 - acc: 0.7236
Epoch 9/100
4074/4074 [==============================] - 1s 313us/sample - loss: 0.5372 - acc: 0.7324
Epoch 10/100
4074/4074 [==============================] - 1s 314us/sample - loss: 

In [334]:
model.evaluate(X_test, y_test)

2000/2000 [==============================] - 0s 122us/sample - loss: 0.4908 - acc: 0.7540


[0.4907910912036896, 0.754]

In [335]:
y_predicted_2 = model.predict(X_test)
y_predicted_2 = np.asanyarray(y_predicted_2)
y_predict_2 = []
for i in range(len(y_predicted_2)):
    if y_predicted_2[i] > 0.5:
        y_predict_2.append(1)
    else:
        y_predict_2.append(0)

In [336]:
print(classification_report(y_test, y_predict_2))

              precision    recall  f1-score   support

           0       0.93      0.74      0.83      1588
           1       0.45      0.79      0.57       412

    accuracy                           0.75      2000
   macro avg       0.69      0.77      0.70      2000
weighted avg       0.83      0.75      0.77      2000



for 3rd class dataset

In [337]:
X_train3.shape, X_test.shape

((4074, 10), (2000, 10))

In [338]:
model = keras.Sequential([
    keras.layers.Dense(15, input_shape = (10, ), activation = 'relu'),
    keras.layers.Dense(10, activation = 'relu'),
    keras.layers.Dense(1,activation='sigmoid')
])

model.compile(
    optimizer = 'adam',
    loss = 'binary_crossentropy', 
    metrics = ['accuracy']
)

model.fit(X_train3, y_train3, epochs=100, batch_size = 10)

Epoch 1/100
4074/4074 [==============================] - 1s 296us/sample - loss: 0.6674 - acc: 0.5987
Epoch 2/100
4074/4074 [==============================] - 1s 236us/sample - loss: 0.6284 - acc: 0.6527
Epoch 3/100
4074/4074 [==============================] - 1s 245us/sample - loss: 0.6005 - acc: 0.6750
Epoch 4/100
4074/4074 [==============================] - 1s 234us/sample - loss: 0.5758 - acc: 0.7069
Epoch 5/100
4074/4074 [==============================] - 1s 277us/sample - loss: 0.5548 - acc: 0.7251
Epoch 6/100
4074/4074 [==============================] - 1s 293us/sample - loss: 0.5419 - acc: 0.7327
Epoch 7/100
4074/4074 [==============================] - 1s 238us/sample - loss: 0.5303 - acc: 0.7440
Epoch 8/100
4074/4074 [==============================] - 1s 243us/sample - loss: 0.5260 - acc: 0.7445
Epoch 9/100
4074/4074 [==============================] - 1s 263us/sample - loss: 0.5191 - acc: 0.7538
Epoch 10/100
4074/4074 [==============================] - 1s 236us/sample - loss: 

In [339]:
model.evaluate(X_test, y_test)

2000/2000 [==============================] - 0s 152us/sample - loss: 0.4235 - acc: 0.7980


[0.4235216660499573, 0.798]

In [340]:
y_predicted_3 = model.predict(X_test)
y_predicted_3 = np.asanyarray(y_predicted_3)
y_predict_3 = []
for i in range(len(y_predicted_3)):
    if y_predicted_3[i] > 0.5:
        y_predict_3.append(1)
    else:
        y_predict_3.append(0)

In [341]:
print(classification_report(y_test, y_predict_3))

              precision    recall  f1-score   support

           0       0.92      0.82      0.87      1588
           1       0.51      0.73      0.60       412

    accuracy                           0.80      2000
   macro avg       0.71      0.77      0.73      2000
weighted avg       0.83      0.80      0.81      2000



for 4th class dataset

In [342]:
X_train4.shape, X_test.shape

((3889, 10), (2000, 10))

In [343]:
model = keras.Sequential([
    keras.layers.Dense(15, input_shape = (10, ), activation = 'relu'),
    keras.layers.Dense(10, activation = 'relu'),
    keras.layers.Dense(1,activation='sigmoid')
])

model.compile(
    optimizer = 'adam',
    loss = 'binary_crossentropy', 
    metrics = ['accuracy']
)

model.fit(X_train4, y_train4, epochs=100, batch_size = 10)

Epoch 1/100
3889/3889 [==============================] - 1s 304us/sample - loss: 0.6627 - acc: 0.5971
Epoch 2/100
3889/3889 [==============================] - 1s 267us/sample - loss: 0.6170 - acc: 0.6583
Epoch 3/100
3889/3889 [==============================] - 1s 232us/sample - loss: 0.5967 - acc: 0.6791
Epoch 4/100
3889/3889 [==============================] - 1s 237us/sample - loss: 0.5792 - acc: 0.6974
Epoch 5/100
3889/3889 [==============================] - 1s 231us/sample - loss: 0.5663 - acc: 0.7118
Epoch 6/100
3889/3889 [==============================] - 1s 231us/sample - loss: 0.5548 - acc: 0.7143
Epoch 7/100
3889/3889 [==============================] - 1s 234us/sample - loss: 0.5453 - acc: 0.7295
Epoch 8/100
3889/3889 [==============================] - 1s 248us/sample - loss: 0.5359 - acc: 0.7331
Epoch 9/100
3889/3889 [==============================] - 1s 234us/sample - loss: 0.5307 - acc: 0.7398
Epoch 10/100
3889/3889 [==============================] - 1s 237us/sample - loss: 

In [344]:
model.evaluate(X_test, y_test)

2000/2000 [==============================] - 0s 153us/sample - loss: 0.4385 - acc: 0.7985


[0.4384655442237854, 0.7985]

In [345]:
y_predicted_4 = model.predict(X_test)
y_predicted_4 = np.asanyarray(y_predicted_4)
y_predict_4 = []
for i in range(len(y_predicted_4)):
    if y_predicted_4[i] > 0.5:
        y_predict_4.append(1)
    else:
        y_predict_4.append(0)

In [346]:
print(classification_report(y_test, y_predict_4))

              precision    recall  f1-score   support

           0       0.92      0.82      0.87      1588
           1       0.51      0.72      0.60       412

    accuracy                           0.80      2000
   macro avg       0.71      0.77      0.73      2000
weighted avg       0.83      0.80      0.81      2000



In [347]:
len(y_predict_1),len(y_predict_2), len(y_predict_3),len(y_predict_4)

(2000, 2000, 2000, 2000)

The majority vote out of all the y_predicted values

As i have even classes, so i am taking the copy of maximum precision and recall class(y_predict_3) and when he count 
is 2 then i am assigning the value as of y_predict_3

In [348]:
y_predict_result = y_predict_3.copy()
for i in range(len(y_predict_4)):
    value = y_predict_1[i] + y_predict_2[i]+ y_predict_3[i] + y_predict_4[i]
    if value <= 1:
        y_predict_result[i] = 0
    if value == 2 :
        y_predict_result[i] = y_predict_3[i]
    if value > 2 :
        y_predict_result[i] = 1
    

Final confusion matrix of ensemble method

In [349]:
confusion_matrix(y_test, y_predict_result)

array([[1302,  286],
       [ 111,  301]], dtype=int64)

Final classification report of ensemble method

In [350]:
print(classification_report(y_test, y_predict_result))

              precision    recall  f1-score   support

           0       0.92      0.82      0.87      1588
           1       0.51      0.73      0.60       412

    accuracy                           0.80      2000
   macro avg       0.72      0.78      0.74      2000
weighted avg       0.84      0.80      0.81      2000



Hence the f1-score of  the classes 0 - (0.92) and 1 - (0.59) has changed to classes 0 - (0.87) and 1 - (0.60)..  
So, overall f1 Score is better than the previous ANN model wihout handling imbalanced dataset